File: C#/Aplicatii client-server/labClientServerCSharpProtoBuf/ClientServer/Protobuff/ProtoWorker.cs
using Google.Protobuf;
using lab3.domain;
using Services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Protobuff
{
    public class ProtoWorker : IObserver
    {
        private IServices server;
        private TcpClient connection;

        private NetworkStream stream;
        private volatile bool connected;
        public ProtoWorker(IServices server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    Request request = Request.Parser.ParseDelimitedFrom(stream);
                    Response response = handleRequest(request);
                    if (response != null)
                    {
                        sendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private Response handleRequest(Request request)
        {
            Response response = null;
            Request.Types.Type reqType = request.Type;
            switch (reqType)
            {
                case Request.Types.Type.Login:
                    {
                        Console.WriteLine("Login request...");
                        Oficiu oficiu = ProtoUtils.getOficiu(request);
                        try
                        {
                            lock (server)
                            {
                                server.login(oficiu, this);
                            }
                            return ProtoUtils.createOkResponse();
                        }
                        catch (MyException ex)
                        {
                            connected = false;
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.GetOficiuByUserPass:
                    {
                        Console.WriteLine("GetOficiuByUserPass request...");
                        string username = ProtoUtils.getUsername(request);
                        string password = ProtoUtils.getPassword(request);
                        try
                        {
                            lock (server)
                            {
                                Oficiu oficiu = server.findOficiuByUserPass(username, password);
                                if (oficiu != null)
                                {
                                    return ProtoUtils.createGetOficiuByUserPassResponse(oficiu);
                                }
                                else
                                {
                                    connected = false;
                                    return ProtoUtils.createErrorResponse("Username si parola gresite!");
                                }
                            }
                        }
                        catch (MyException ex)
                        {
                            connected = false;
                            return ProtoUtils.createErrorResponse("Username si parola gresite!");
                        }
                    }
                case Request.Types.Type.Logout:
                    {
                        Console.WriteLine("Logout request...");
                        Oficiu oficiu = ProtoUtils.getOficiu(request);
                        try
                        {
                            lock (server)
                            {
                                server.logout(oficiu, this);
                            }
                            connected = false;
                            return ProtoUtils.createOkResponse();
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.GetAllParticipants:
                    {
                        Console.WriteLine("GetAllParticipant request...");
                        try
                        {
                            lock (server)
                            {
                                lab3.domain.Participant[] participanti = server.findAllParticipant();
                                return ProtoUtils.createGetAllParticipantResponse(participanti);
                            }
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.GetAllCursa:
                    {
                        Console.WriteLine("GetAllCursa request...");
                        try
                        {
                            lock (server)
                            {
                                lab3.domain.Cursa[] curse = server.findAllCursa();
                                return ProtoUtils.createGetAllCursaResponse(curse);
                            }
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.GetAllParticipantsByTeam:
                    {
                        Console.WriteLine("FindAllParticipantEchipa request...");
                        string echipa = ProtoUtils.getEchipa(request);
                        try
                        {
                            lock (server)
                            {
                                lab3.domain.Participant[] participanti = server.findAllParticipant(echipa);
                                return ProtoUtils.createGetAllParticipantResponse(participanti);
                            }
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.GetEchipe:
                    {
                        Console.WriteLine("GetEchipe request...");
                        try
                        {
                            lock (server)
                            {
                                string echipe = "";
                                return ProtoUtils.createGetEchipeResponse(echipe);
                            }
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.GetParticipant:
                    {
                        Console.WriteLine("GetParticipant request...");
                        string nume = ProtoUtils.getNume(request);
                        string echipa = ProtoUtils.getEchipa(request);
                        int capMotor = ProtoUtils.getCapMotor(request);
                        try
                        {
                            lock (server)
                            {
                                lab3.domain.Participant participant = server.findParticipant(nume, echipa, capMotor);
                                return ProtoUtils.createGetParticipantResponse(participant);
                            }
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }

                case Request.Types.Type.GetCursa:
                    {
                        Console.WriteLine("GetCursa request...");
                        string id = ProtoUtils.getIdCursa(request);
                        try
                        {
                            lock (server)
                            {
                                lab3.domain.Cursa cursa = server.findCursa(id);
                                return ProtoUtils.createGetCursaResponse(cursa);
                            }
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }

                case Request.Types.Type.AddParticipantCursa:
                    {
                        Console.WriteLine("AddParticipantCursa request...");
                        lab3.domain.ParticipantCursa participantCursa = ProtoUtils.getParticipantCursa(request);
                        try
                        {
                            lock (server)
                            {
                                server.saveParticipantCursa(participantCursa);
                            }
                            return ProtoUtils.createOkResponse();
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
                case Request.Types.Type.UpdateCursa:
                    {
                        Console.WriteLine("UpdateCursa request...");
                        string idCursa = ProtoUtils.getIdCursa(request);
                        lab3.domain.Cursa cursa = ProtoUtils.getCursa(request);
                        cursa.Id = idCursa;
                        try
                        {
                            lock (server)
                            {
                                server.deleteCursa(idCursa);
                                server.saveCursa(cursa);
                                server.submitted(server.findAllCursa());
                            }
                            return ProtoUtils.createOkResponse();
                        }
                        catch (MyException ex)
                        {
                            return ProtoUtils.createErrorResponse(ex.Message);
                        }
                    }
            }
            return response;
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            lock (stream)
            {
                response.WriteDelimitedTo(stream);
                stream.Flush();
            }
        }

        public virtual void participantCursaAdded(lab3.domain.Cursa[] curse)
        {
            Console.WriteLine("New participantCursa added...");
            try
            {
                sendResponse(ProtoUtils.createNewParticipantCursaAddedResponse());
            }
            catch (Exception e)
            {
                throw new MyException("Sending error " + e);
            }
        }
    }
}

File: Java/Aplicatii client-server/labJavaClientServerProtoBuf/Networking/src/main/java/ProtocolBuffers/ProtoWorker.java
package ProtocolBuffers;

import dto.CursaDTO;
import dto.DTOUtils;
import dto.OficiuDTO;
import dto.ParticipantDTO;
import model.Cursa;
import model.Oficiu;
import model.Participant;
import model.ParticipantCursa;
import objectProtocol.*;
import services.IObserver;
import services.IServices;
import services.MyException;

import java.io.*;
import java.net.Socket;

public class ProtoWorker implements Runnable, IObserver {
    private IServices server;
    private Socket connection;

    //private ObjectInputStream input;
    //private ObjectOutputStream output;
    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public ProtoWorker(IServices server, Socket connection){
        this.server = server;
        this.connection = connection;
        try{
            output = connection.getOutputStream(); //new ObjectOutputStream(connection.getOutputStream());
            input = connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    @Override
    public void run() {
        while(connected){
            try{
                //Object request = input.readObject();
                Protocol.Request request = Protocol.Request.parseDelimitedFrom(input);
                Protocol.Response response = handleRequest(request);
                if(response != null){
                    sendResponse(response);
                }
            } catch (IOException | MyException e) {
                e.printStackTrace();
            }
            try{
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try{
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error " + e);
        }
    }

    public void participantCursaAdded() throws MyException{
        System.out.println("ParticipantCursa added");
        try{
            sendResponse(ProtoUtils.createNewParticipantCursaAddedResponse());
        }
        catch (IOException e){
            throw new MyException("Sending error: " + e);
        }
    }

    private void sendResponse(Protocol.Response response) throws IOException{
        System.out.println("Sending response..." + response);
        //output.writeObject(response);
        response.writeDelimitedTo(output);
        output.flush();
    }

    private Protocol.Response handleRequest(Protocol.Request request) throws MyException {
        Protocol.Response response = null;
        switch (request.getType()){
            case Login:{
                System.out.println("Login request...");
                Oficiu oficiu = ProtoUtils.getOficiu(request);
                try{
                    server.findOficiuByUserPass(oficiu.getUsername(), oficiu.getPassword(), this);
                    return ProtoUtils.createOkReponse();
                } catch (MyException e) {
                    connected = false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetOficiuByUserPass:{
                System.out.println("GetOficiuByUserPass request");
                String username = ProtoUtils.getUsername(request);
                String password = ProtoUtils.getPassword(request);
                Oficiu oficiu = server.findOficiuByUserPass(username, password);
                return ProtoUtils.createGetOficiuByUserPassResponse(oficiu);
            }
            case Logout:{
                System.out.println("Logout request");
                Oficiu oficiu = ProtoUtils.getOficiu(request);
                System.out.println("Oficiul in ClientWorker este: " + oficiu);
                try{
                    server.logout(oficiu, this);
                    connected = false;
                    return ProtoUtils.createOkReponse();
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetAllParticipants:{
                System.out.println("GetAllParticipants request");
                try{
                    Participant[] participants = server.findAllParticipant();
                    return ProtoUtils.createGetAllParticipantsResponse(participants);
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetAllCursa:{
                System.out.println("GetAllCursa request");
                try{
                    Cursa[] curse = server.findAllCursa();
                    return ProtoUtils.createGetAllCursaResponse(curse);
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetAllParticipantsByTeam:{
                System.out.println("GetAllParticipantsByTeam request");
                String echipa = ProtoUtils.getEchipa(request);
                try{
                    Participant[] participants = server.findAllParticipant(echipa);
                    return ProtoUtils.createGetAllParticipantsResponse(participants);
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetEchipe:{
                System.out.println("GetEchipe request");
                try{
                    String echipe = server.findAllEchipe();
                    return ProtoUtils.createGetEchipeResponse(echipe);
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetParticipant:{
                System.out.println("GetParticipant request");
                String nume = ProtoUtils.getNume(request);
                String echipa = ProtoUtils.getEchipa(request);
                Integer capMotor = ProtoUtils.getCapMotor(request);
                try{
                    Participant participant = server.findParticipant(nume, echipa, capMotor);
                    return ProtoUtils.createGetParticipantResponse(participant);
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetCursa:{
                System.out.println("GetCursa request");
                String idCursa = ProtoUtils.getIdCursa(request);
                try{
                    Cursa cursa = server.findCursa(idCursa);
                    return ProtoUtils.createGetCursaResponse(cursa);
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case AddParticipantCursa:{
                System.out.println("AddParticipantCursa request");
                //ParticipantCursaDTO participantCursaDTO = req.getParticipantCursaDTO();
                //System.out.println("Dto: ");
                //System.out.println(participantCursaDTO);
                //ParticipantCursa participantCursa = DTOUtils.getFromDTO(participantCursaDTO);
                ParticipantCursa participantCursa = ProtoUtils.getParticipantCursa(request);
                try{
                    System.out.println("Participantul Cursa de salvat este:");
                    System.out.println(participantCursa);
                    server.saveParticipantCursa(participantCursa);
                    return ProtoUtils.createOkReponse();
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case UpdateCursa:{
                System.out.println("UpdateCursa request");
                String idCursa = ProtoUtils.getIdCursa(request);
                Cursa cursa = ProtoUtils.getCursa(request);
                try{
                    server.updateCursa(idCursa, cursa);
                    return ProtoUtils.createOkReponse();
                } catch (MyException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
        }
        return response;
    }
}
